contrib/epee/include/syncobj.h

```diff
@@ -150,6 +154,127 @@ namespace epee
   };
 
 
+  struct reader_writer_lock {
+
+    reader_writer_lock() noexcept : 
+    write_mode(false),
+    read_mode(0) {};
+
+    bool start_read() noexcept {
+      if (have_write()) {
+        return false;
+      }
+      lock_reader();
+      return true;
+    }
+
+    void end_read() noexcept {
+      unlock_reader();
+      condition.notify_all();
+    }
+
+    bool start_write() noexcept {
+      if (have_write()) {
+        return false;
+      }
+      lock_and_change(boost::this_thread::get_id());
+      return true;
+    }
+
+    void end_write() noexcept {
+      unlock_writer();
+      condition.notify_all();
+    }
+
+    bool have_write() noexcept {
+      boost::mutex::scoped_lock lock(internal_mutex);
+      if (write_mode
+          && writer_id == boost::this_thread::get_id()) {
+        return true;
+      }
+      return false;
+    }
+
+    ~reader_writer_lock() = default;
+    reader_writer_lock(reader_writer_lock&) = delete;
+    reader_writer_lock operator=(reader_writer_lock&) = delete;
+
+  private:
+
+    void unlock_writer() noexcept {
+      CHECK_AND_ASSERT_MES2(write_mode && !read_mode, "Ending write transaction by " << boost::this_thread::get_id() << " while the lock is not in write mode");
+      boost::mutex::scoped_lock lock(internal_mutex);      
+      rw_mutex.unlock();
+      write_mode = false;
+      writer_id = boost::thread::id();
+    }
+
+    void unlock_reader() noexcept {
+      CHECK_AND_ASSERT_MES2(read_mode && !write_mode, "Ending read transaction by " << boost::this_thread::get_id() << " while the lock is not in read mode");
+      boost::mutex::scoped_lock lock(internal_mutex);      
+      rw_mutex.unlock_shared();
+      read_mode--;
+      if (rw_mutex.try_lock()) {
+        CHECK_AND_ASSERT_MES2(!read_mode, "Reader is not zero but goes to read_mode = 0 by " << boost::this_thread::get_id());
+        write_mode = false;
+        rw_mutex.unlock();
+      }
+      return; 
+    }
+
+    void lock_reader() noexcept {
+      CHECK_AND_ASSERT_MES2(read_mode < UINT32_MAX, "Maximum number of readers reached.");
+      do {  
+        boost::mutex::scoped_lock lock(internal_mutex);      
+        if(!write_mode && rw_mutex.try_lock_shared()) {
+          write_mode = false;
+          read_mode++;
+          return; 
+        }
+        condition.wait(lock);
+      } while(true);
+    }
+
+    void lock_and_change(boost::thread::id new_owner) noexcept {
+      do {  
+        boost::mutex::scoped_lock lock(internal_mutex);      
+        if (!read_mode && rw_mutex.try_lock()) {
+          writer_id = new_owner;
+          write_mode = true;
+          read_mode = 0;
+          return; 
+        }
+        condition.wait(lock);
+      } while(true);
+    }
```

Comment:
Unfortunately, this locking/waiting pattern *may* cause more issues than the current code because we are now much more likely to experience writer starvation. Before, if there was 1000 readers and one writer, the writer was on equal footing as the rest of the write threads. Current code is still not fair, but the writer thread has the same chance as acquiring exclusive access as all other threads, which are attempting to acquire exclusively as well. However, now, the write lock will continually thrash and refuse to even attempt the lock if there is even one active reader, indefinitely. As long as new readers keep coming in, the writer is never even guaranteed the chance to even ask for the lock. In practice, this may cause the chain to never move forward with enough readers.

Reply:
*Starvation* might happen in extremely rare cases. That is why I am insisting this is an **_intermediary_** step until most of the APIs inside `Blockchain` are fixed. 

But your analysis is **_not correct_**. In the current locking mechanism, we lock the blockchain for every read. Let me explain a hypothetical scenario: 

There are multiple read transactions: `R1`, `R2`, `R3`, `R4`, `R5`, and `R6`.
There are single write transactions: `W1`. 

In the current locking mechanism: 
`R1` -> `R2`  -> `R3`  -> `R4`  -> `R5`  -> `R6` -> `W1`

But in reader-writer locking mechanism:


Core1: `R1` `R2`  -> `R3`  -> `R4`  -> `R5`  -> `R6`
Core1: 
Core1: 
Core1: 
Core1: 
Core1: 
Core1: 
Core1:

Reply:
*Starvation* might happen in extremely rare cases. That is why I am insisting this is an **_intermediary_** step until most of the APIs inside `Blockchain` are fixed. 

But your analysis is **_not correct_**. In the current locking mechanism, we lock the entire blockchain for every read. Let me explain a hypothetical scenario: 

There are multiple read transactions: `R1`, `R2`, `R3`, `R4`, `R5`, `R6`, `R7`, and `R8`.
There are single write transactions: `W1`. 

In the current locking mechanism: 
`Core1`: `R1` -> `R2`  -> `R3`  -> `R4`  -> `R5`  -> `R6` -> `R7`  -> `R8` -> `W1`
`Core2`:  
`Core3`: 
`Core4`: 
`Core5`: 
`Core6`: 
`Core7`: 
`Core8`: 

But in reader-writer locking mechanism:
`Core1`: `R1` `W1`
`Core2`: `R2` 
`Core3`: `R3`
`Core4`: `R4`
`Core5`: `R5`
`Core6`: `R6`
`Core7`: `R7`
`Core8`: `R8`



Now what you saying is a stream of `read` transactions coming to starve the `write`. It is a *theoretical* possibility.

Reply:
McKenney, Paul E., Silas Boyd-Wickizer, and Jonathan Walpole. "RCU usage in the linux kernel: One decade later." Technical report (2013).

Exactly talks about this issue in the section `5.3 Retry Readers`.

Reply:
It was fixed in a new push. The `reader_writer` lock does have an internal `queue` right now. The `queue` is for keeping track of work. 

Let me give you an example of how it works now: 


There are multiple read transactions: `R1`, `R2`, `R3`, `R4`, `R5`, `R6`, `R7`, `R8`, `R9`, `R10`, `R11`.
There are single write transactions: `W1`. 

The order of arrival is: `R1`, `R2`, `R3`, `R4`, `R5`, `R6`, `R7`, `R8`, `W1`, `R9`, `R10`, `R11`.

In the current locking mechanism: 
`Core1`: `R1` -> `R2`  -> `R3`  -> `R4`  -> `R5`  -> `R6` -> `R7`  -> `R8` -> `W1` -> `R9`-> `R10`-> `R11` 
`Core2`:  
`Core3`: 
`Core4`: 
`Core5`: 
`Core6`: 
`Core7`: 
`Core8`: 

But in new reader-writer locking mechanism:
`Core1`: `R1` `W1` `R9`
`Core2`: `R2`          `R10`
`Core3`: `R3`          `R11`
`Core4`: `R4`
`Core5`: `R5`
`Core6`: `R6`
`Core7`: `R7`
`Core8`: `R8`